Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/dialogController/MenuController.java

```diff
@@ -80,159 +84,120 @@ public void onEvent(Event event) throws Exception {
-                getBaseListboxController().cut();
-              }
-            });
-            fileMenupopup.insertBefore(item, targetMenuitem);
-
-            item = new Menuitem();
-            item.setLabel(Labels.getLabel("common_copy_text"));
-            item.setImage("~./themes/ap/common/img/icons/copy.svg");
-            item.addEventListener(ON_CLICK, new EventListener<Event>() {
-              @Override
-              public void onEvent(Event event) throws Exception {
-                getBaseListboxController().copy();
-              }
-            });
-            fileMenupopup.insertBefore(item, targetMenuitem);
-
-            item = new Menuitem();
-            item.setLabel(Labels.getLabel("common_paste_text"));
-            item.setImage("~./themes/ap/common/img/icons/paste.svg");
-            item.addEventListener(ON_CLICK, new EventListener<Event>() {
-              @Override
-              public void onEvent(Event event) throws Exception {
-                getBaseListboxController().paste();
-              }
-            });
-            fileMenupopup.insertBefore(item, targetMenuitem);
+    // If there are portal plugins, create the menus for launching them
+    if (!PortalPluginResolver.resolve().isEmpty()) {
+      portalPluginMap = PortalPluginResolver.getPortalPluginMap();
 
-            sep = new Menuseparator();
-            fileMenupopup.insertBefore(sep, targetMenuitem);
+      List<String> menuGroups = menuConfig.getMenuGroups();
+      Map<String, List<String>> menuiItemMap = menuConfig.getMenuiItemMap();
 
-            if (etlMenuItem != null) {
-              sep = new Menuseparator();
-              fileMenupopup.insertBefore(sep, etlMenuItem);
+      for (String menuGroupId : menuGroups) {
+        List<String> itemIds = menuiItemMap.get(menuGroupId);
+        if (itemIds.size() == 0) {
+          continue;
+        }
+        String groupLabelKey = MessageFormat.format("plugin_{0}_title_text", menuGroupId.toLowerCase());
+        String groupLabel = Labels.getLabel(groupLabelKey);
+        Menu menu = new Menu(groupLabel);
+        menu.setClientDataAttribute(GROUP, menuGroupId);
+        Menupopup popup = new Menupopup();
+        for (String itemId : itemIds) {
+          if (Constants.PLUGIN_CUT.equals(itemId)) {
+            addCutMenuItem(popup);
+          } else if (Constants.PLUGIN_COPY.equals(itemId)) {
+            addCopyMenuItem(popup);
+          } else if (Constants.PLUGIN_PASTE.equals(itemId)) {
+            addPasteMenuItem(popup);
+          } else if (Constants.ITEM_SEPARATOR.equals(itemId)) {
+            if (!(popup.getLastChild() instanceof Menuseparator)) {
+              Menuseparator separator = new Menuseparator();
+              popup.appendChild(separator);
             }
-          } catch (Exception e) {
-            LOGGER.error("Ignored exception during main menu construction", e);
+          } else {
+            addPluginToMenu(popup, itemId);
           }
         }
```

Comment:
The Cut, Copy, Paste can be managed consistently under the same PortalPluginResolver mechinism (via PortalPlugin interface and beans). They are different from other PortalPlugin such as PD or Calendar in the sense that they are _internal_ PortalPlugin, i.e. their action could be handled internally by classes within the Portal module itself. 

Suggest to create a package called plugins inside the Portal, then create one plugin class for each editing action, e.g. CutPortalPlugin, CopyPortalPlugin, PastePortalPlugin which implements PortalPlugin interface, and implement the current interface such as execute. They can be loaded via PortalPluginResolver, no need to create conditionals/exceptional cases for these plugins which add another layer of complexity to manage.

Reply:
I will implement the internal plugins for cut, copy, paste in a separate ticket